src/python_pachyderm/mixin/pps.py

```diff
@@ -201,6 +204,87 @@ def create_pipeline(self, pipeline_name, transform, parallelism_spec=None, hasht
         * `sidecar_resource_limits`: An optional `ResourceSpec` setting
         resource limits for the pipeline sidecar.
         """
+
+        # Support for build step-enabled pipelines. This is a python port of
+        # the equivalent functionality in pachyderm core's
+        # 'src/server/pps/cmds/cmds.go', and any changes made here likely have
+        # to be reflected there as well.
+        if transform.build.image or transform.build.language or transform.build.path:
+            if spout:
+                raise Exception("build step-enabled pipelines do not work with spouts")
+            if not input:
+                raise Exception("no `input` specified")
+            if (not transform.build.language) and (not transform.build.image):
+                raise Exception("must specify either a build `language` or `image`")
+            if transform.build.language and transform.build.image:
+                raise Exception("cannot specify both a build `language` and `image`")
+            if any(pipeline_input_name(i) in ("build", "source") for i in pipeline_inputs(input)):
+                raise Exception(
+                    "build step-enabled pipelines cannot have inputs with the name "
+                    + "'build' or 'source', as they are reserved for build assets"
+                )
+
+            build_path = Path(transform.build.path or ".")
+            if not build_path.exists():
+                raise Exception("build path {} does not exist".format(build_path))
+            if (build_path / ".pachignore").exists():
+                warnings.warn(
+                    "detected a '.pachignore' file, but it's unsupported by python_pachyderm -- use `pachctl` instead",
+                    RuntimeWarning
+                )
+
+            build_pipeline_name = "{}_build".format(pipeline_name)
+
+            image = transform.build.image
+            if not image:
+                version = self.get_remote_version()
+                version_str = "{}.{}.{}{}".format(version.major, version.minor, version.micro, version.additional)
+                image = "pachyderm/{}-build:{}".format(transform.build.language, version_str)
+            if not transform.image:
+                transform.image = image
+
+            def create_build_pipeline_input(name):
+                return pps_proto.Input(
+                    pfs=pps_proto.PFSInput(
+                        name=name,
+                        glob="/",
+                        repo=build_pipeline_name,
+                        branch=name,
+                    )
+                )
+
+            self.create_repo(build_pipeline_name, update=True)
```

Comment:
1) With build pipelines (but unlike the version of `create_python_pipeline` that this is replacing), there is only the `_build` repo, which contains the source code in one branch, and the build assets pipeline outputting to another branch.
2) I'm not sure if I follow your second question. Is it "why are we calling `self._req` directly rather than `self.create_pipeline` in the following line?"

Reply:
I think I just didn't quite understand how this feature worked (that the source and build output were two branches in the same repo), but this clarifies it—thanks!